docs: Add a setup documentation about examples/kv-cache-index (#38)
llm-d/llm-d-kv-cache-manager: docs/deployment/setup.md

# KV-Cache Manager Setup Guide

This guide provides a complete walkthrough for setting up and testing the example llm-d-kv-cache-manager system. You will deploy a vLLM with LMCache and Redis, then run an example application that demonstrates KV cache indexing capabilities.

By following this guide, you will:

> 1. **Deploy the Infrastructure**: Use Helm to set up a Kubernetes cluster running:

Reviewer: Technically the Kubernetes cluster is a pre-requisite and is not being set up here. This is well reflected below, perhaps the postfix sentence can just be dropped.

>    - vLLM servers (default: 4 replicas) serving Llama 3.1 8B Instruct model,

Reviewer: I think commas should be removed from list items.

   - Redis server,
   - LMCache,
2. **Test with Example Application**: Run a Go application that:
   - Connects to your deployed vLLM and Redis infrastructure,
   - Demonstrates KV cache indexing by processing a sample prompt,
   - Shows how to retrieve pod scores, 

The end result is a working distributed LLM system with intelligent KV cache management that can route requests to pods with relevant cached computations.

## vLLM Deployment

> The llm-d-kv-cache-manager repository includes a Helm chart for deploying vLLM with LMCache and Redis. This section describes how to use this Helm chart for a complete deployment.

Reviewer: s/vLLM with LMCache and Redis/vLLM with CPU offloading (LMCache) and KV-events indexing (Redis)/.

*Note*: Ensure that the Kubernetes node designated for running vLLM supports GPU workloads.

### Prerequisites

- Kubernetes cluster with GPU support
- Helm 3.x
- HuggingFace token for accessing models
- kubectl configured to access your cluster

### Installation

1. Set environment variables:

```bash
export HF_TOKEN=<your-huggingface-token>
export NAMESPACE=<your-namespace>
export MODEL_NAME="meta-llama/Llama-3.1-8B-Instruct"
```

> Note that both the Helm deployment and the example application use the same `MODEL_NAME` environment variable,
> ensuring alignment between the vLLM deployment configuration and the KV cache indexer.
> Set this variable once during initial setup and both components will use the same model configuration.

2. Deploy using Helm:

```bash
helm upgrade --install vllm-stack ./vllm-setup-helm \
  --namespace $NAMESPACE \
  --create-namespace \
  --set secret.create=true \
  --set secret.hfTokenValue=$HF_TOKEN \
  --set vllm.model.name=$MODEL_NAME \

>   --set vllm.poolLabelValue="vllm-model-pool" \

Reviewer: Think it would be clearer if this is also an environment variable.

# KV-Cache Manager Setup Guide

This guide provides a complete walkthrough for setting up and testing the example llm-d-kv-cache-manager system. You will deploy a vLLM with LMCache and Redis, then run an example application that demonstrates KV cache indexing capabilities.

By following this guide, you will:

1. **Deploy the Infrastructure**: Use Helm to set up:

>    - vLLM servers (default: 4 replicas) serving Llama 3.1 8B Instruct model

Reviewer: ```suggestion
   - vLLM nodes with LMCache CPU offloading (4 replicas) serving Llama 3.1 8B Instruct model
```

Apologies for missing earlier, LMCache being a standalone item makes it sound like it will run separately. The LMCache engine is initialized within the vLLM processes.

Author: Thank you for pointing out this! 🙏 i have updated the code

   - Redis server

>    - LMCache

Reviewer: ```suggestion
```

2. **Test with Example Application**: Run a Go application that:
   - Connects to your deployed vLLM and Redis infrastructure,

>    - Demonstrates KV cache indexing by processing a sample prompt
>    - Shows how to retrieve pod scores

Reviewer: ```suggestion
   - Demonstrates KV cache indexing by processing a sample prompt
```

The concept of scoring here can be confusing.

> The end result is a working distributed LLM system with intelligent KV cache management that can route requests to pods with relevant cached computations.

Reviewer: ```suggestion
The demonstrated KV-cache indexer is utilized for AI-aware routing to accelerate inference across the system through minimizing redundant computation.
```

This line would be true for a demo with a router. This setup guide only runs the indexer.

- By default, the chart uses a `PersistentVolume` to cache the model. To disable this, set `.persistence.enabled` to `false`.

3. Verify the deployment:

```bash
kubectl get deployments -n $NAMESPACE
```

You should see:

- vLLM pods (default: 4 replicas)
- Redis lookup server pod

### Configuration Options

The Helm chart supports various configuration options. See [values.yaml](../../vllm-setup-helm/values.yaml) for all available options.

Key configuration parameters:

- `vllm.model.name`: The HuggingFace model to use (default: `meta-llama/Llama-3.1-8B-Instruct`)
- `vllm.replicaCount`: Number of vLLM replicas (default: 4)
- `vllm.poolLabelValue`: Label value for the inference pool (used by scheduler)
- `redis.enabled`: Whether to deploy Redis for KV cache indexing (default: true)
- `persistence.enabled`: Enable persistent storage for model cache (default: true)
- `secret.create`: Create HuggingFace token secret (default: true)

## Using the KV Cache Indexer Example

### Prerequisites

Ensure you have a running deployment with vLLM and Redis as described above.

### Running the Example

The vLLM node can be tested with the prompt found in `examples/kv-cache-index/main.go`.

> First, ensure that the tokenizer engine is available locally for our example:

Reviewer: ```suggestion
First, download the tokenizer bindings required by the `kvcache.Indexer` for prompt tokenization:
```

```bash
make download-tokenizer
```

> This will download the tokenizer engine.

Reviewer: ```suggestion
Then, set the required environment variables and run example:
```

# KVCache Manager Setup Guide

> This guide provides instructions for deploying the llm-d-kv-cache-manager with vLLM, LMCache, and Redis using Helm.

Reviewer: I think this opener should better introduce the scope and agenda of this doc. It should describe what is going to be ran in short, including the example and its content.

## Deployment with Helm

The llm-d-kv-cache-manager repository includes a Helm chart for deploying vLLM with LMCache and Redis. This section describes how to use this Helm chart for a complete deployment.

### Prerequisites

- Kubernetes cluster with GPU support
- Helm 3.x
- HuggingFace token for accessing models
- kubectl configured to access your cluster

### Installation

1. Set environment variables:

```bash
export HF_TOKEN=<your-huggingface-token>
export NAMESPACE=<your-namespace>
```

2. Deploy using Helm:

```bash

> helm upgrade --install vllm-p2p ./vllm-setup-helm \

Reviewer: s/vllm-p2p/llm-d-kv-cache-manager/ or any other name you like. At the time p2p referred to p2p sharing, which is not relevant here.

Author: thanks for the context! updated according to your suggestion; `llm-d-kv-cache-manager`
